CSCFAIRMETA-201: Old Etsin redirect, add server block

MattiasLevlin opened the pull request:
- etsin.avointiede.fi must be redirected to etsin.fairdata.fi
- Added server bloc for etsin.avointiede.fi. Secret files have been added (.crt / .key) too.

MattiasLevlin committed:
CSCFAIRMETA-201: Old Etsin redirect, add server block

- etsin.avointiede.fi must be redirected to etsin.fairdata.fi
- Added server bloc for etsin.avointiede.fi. Secret files to be added too.

MattiasLevlin commented:
Certificates have been added to Spacewalk. However, I'm not completely sure about the syntax, so a review would probably be helpful.

hannu40k commented on ansible/roles/nginx/templates/etsin_nginx.conf:
```diff
@@ -193,4 +193,14 @@ http {
+        ssl_certificate_key {{ ssl_certificates_path }}/etsin.avointiede.fi.key;
+
+        return 301 https://etsin.fairdata.fi$request_uri;
+    }
```
Did you test if this redirect works and the used ssl-related parameters are enough? In the other ssl-server blocks, there are quite a few more parameters used.

Otherwise the block looks like it could certainly work. It could be improved a lot by parameterizing server_name, the cert names, and the redirect url. This way the server block can be easer tested in all envs:
- server_name can be for example redirect.etsin.csc.local (or whatever the local dev url pattern is for etsin).
- Similarly for test, stable etc envs, for example redirect.etsin-test.fairdata.fi. Certs for that url should use self-signed certifiactes (created by ansible, in the role "certificates") for other envs than production.
- The name of the used certificate should be parameterized (e.g. {{ old_etsin_cert_name }}, etc)
- The final redirect url should use the existing {{ server_domain_name }} in all envs, so that for example redirect.etsin-test.fairdata.fi redirects to etsin-test.fairdata.fi.

Doing these things is not a lot of work, and should be a good exercise in modifying ansible playbooks and conf files.

MattiasLevlin replied:
Thanks for a thorough review and related instructions. I will improve and test the code based on these change requests.

MattiasLevlin committed:
CSCFAIRMETA-201: Add certificate creation instructions, parameterize variables, add local variables

- Needed in order to follow best practices and parameterize things for different environments
- Certificate creation instructions added in certificates/task/main.yml, added local variables in local_development/group_vars/all, parameterized things in etsin_nginx.conf
- Need to update secret files in spacewalk.csc.fi, as well

hannu40k commented on ansible/roles/certificates/tasks/main.yml:
```diff
@@ -6,6 +6,10 @@
 
+- name: Create self-signed redirect SSL cert and private key
+  command: openssl req -x509 -nodes -subj "/C=FI/ST=Uusimaa/L=Espoo/O=CSC/CN={{ redirect_server_domain_name }}" -days 365 -newkey rsa:2048 -keyout {{ ssl_certificates_path }}/{{ redirect_ssl_certificate_key_name }} -out {{ ssl_certificates_path }}/{{ redirect_ssl_certificate_name }} creates={{ ssl_certificates_path }}/{{ redirect_ssl_certificate_name }}
+  when: deployment_environment_id in ['local_development', 'test', 'stable', 'demo']
```
I think the rule could be ``when: deployment_environment_id not in ['production']``

MattiasLevlin replied:
Thanks. This is a better alternative.

hannu40k commented on ansible/inventories/local_development/group_vars/all:
```diff
@@ -7,16 +7,20 @@ dataserver_1_internal_ip: 127.0.0.1
 gunicorn_preload: false
 
 ssl_certificates_path: /etc/nginx/ssl_certs
+redirect_ssl_certificates_path: /etc/nginx/ssl_certs
```
I think this variable is unnecessary. The existing ``ssl_certificates_path`` directory should be ok for these certs as well.

hannu40k replied:
Actually looks like you did use ``ssl_certificates_path``, but probably just forgot this variable here.

MattiasLevlin replied:
Yes, you are correct, this variable is unnecessary. It was created to account for a separate certificate location for the old production certificate, but since the variable isn't used in production, it's not needed.

hannu40k commented on ansible/inventories/local_development/group_vars/all:
```diff
@@ -7,16 +7,20 @@ dataserver_1_internal_ip: 127.0.0.1
 web_app_github_repo_branch: test
 search_app_github_repo_branch: test
 server_domain_name: etsin-finder.local
+redirect_server_domain_name: redirect.etsin-finder.local
```
Remember to use existing variables when possible. Such as here: ``redirect_server_domain_name: "redirect.{{ server_domain_name}}"``

MattiasLevlin replied:
Thanks, a good practice to follow.

MattiasLevlin committed:
CSCFAIRMETA-201: Code improvements

- Removed unnecessary variable. Reused another variable. Changed a rule to run when not in production.
- Small code changes, as advised by Hannu
- No side effects. However, to be tested still.

MattiasLevlin committed:
CSCFAIRMETA-201: Add new file for shared ssl configurations, new server block for port 80

- Shared SSL configs are needed both in the standard server block and in the redirect. Port 80 is useful in addition to 443
- Moved variables to new file, and import variables at correct places. Add bloc with port 80 to redirect.
- No side effects, but should be tested.

MattiasLevlin committed:
CSCFAIRMETA-201: Add copy shared_ssl_configurations.conf to nginx tasks

- The new file must be copied to etc/nginx... so that it is available on the servers
- Added a block similar to "Copy shared_headers.conf"

MattiasLevlin committed:
CSCFAIRMETA-201: Simplify listen 80 syntax

- Simplified, as there is probably no need for two server blocks
- Using the same block, according to http://nginx.org/en/docs/http/configuring_https_servers.html#single_http_https_server
